### DISCUSSION TITLE
Netbox 3.5 - script storage across multiple containers

### DISCUSSION THREAD

**Discussion by FliesLikeABrick** [netbox-community/netbox#12643] [Q&A]
Hello,

We have multiple containers running on different hosts for HA to our netbox setup.  We previously managed our custom script deployment by tracking those files in our internal fork of netbox-docker, and exposing them to the container via a volume mapping from the local host (that had our netbox-docker fork cloned locally).

With 3.5's change that requires scripts to be uploaded via the UI, it does not look like this strategy works?    Is there a recommended approach to synchronizing these reasonably?  I would like to avoid needing a shared nfs volume or something for the script storage that the UI manages, if possible - since we already have our scripts versioned in git.  Is there a way to just deploy these either in the container build or in the filesystem of the physical host?

**Comment by zmaster7** [comment]
Check out my response in #12663, which would be along the same lines. You'd be able to tie directly into your git repo and sync changes, but as for now the "upload" process is manual, even when a data source is involved. However, the file changes will be synced automatically when checked.

**Comment by abhi1693** [comment]
Have you tried using a local data source? If you shipped everything within the image, you could simply use the local backend to fetch your reports and scripts. It's just an additional step but nothing is required to be changed in your deployment. This also doesn't create anything extra on the filesystem.

**Reply by rasanentimo** [reply]
@abhi1693 I tested this approach as well but it didn't work. I have the scripts locally on path `/scripts` on 2 pods on my test kubernetes setup.


```
unit@netbox-75ffc7db75-476bt:/opt/netbox/netbox$ ls /scripts/
script1.py  script2.py	script3.py  utils
```

and 

```
unit@netbox-75ffc7db75-t5bxt:/opt/netbox/netbox$ ls /scripts
script1.py  script2.py	script3.py  utils
```

When I add the script 'Customization -> Scripts -> Add', the selected files are copied to the scripts path

```
unit@netbox-75ffc7db75-476bt:/opt/netbox/netbox$ ls /opt/netbox/netbox/scripts/
__init__.py  __pycache__  script1.py  script2.py 
```

but these files are not copied on the other container

```
unit@netbox-75ffc7db75-t5bxt:/opt/netbox/netbox$ ls scripts/
__init__.py
```

so if the user ends on pod netbox-75ffc7db75-t5bxt, the scripts are not visible. The whole process would need to be done again if the pods are deleted.

![image](https://github.com/netbox-community/netbox/assets/10869276/8edc5100-45b5-42a9-8cb9-c4a6a7549fc6)


Let me know if I missed some steps to make this work.

**Reply by abhi1693** [reply]
Then to complete this process you would probably need a volume that is shared between your pods for your script's root path where these files are copied. You may also need to ensure the volume supports multiple read-write from all these pods.

**Reply by rasanentimo** [reply]
I guess that would the only option. However, I would prefer to have an option to load the custom scripts, reports etc same way as prior 3.5 release. This would allow to keep the Netbox app stateless, use read only file system, avoid any manual actions etc.

**Reply by abhi1693** [reply]
To keep it stateless, you can help fix this bug: https://github.com/netbox-community/netbox/issues/13095

**Reply by rasanentimo** [reply]
Can that be used also scripts/reports which are originated from local/git data source or is this only for manually uploaded files? 

Having s3fs mount for the scripts directory could also work.

**Reply by abhi1693** [reply]
The idea of the fix is to keep the reports and scripts on S3 (if defined) irrespective of how they were synced

**Reply by rasanentimo** [reply]
ok, that sounds good! However, I guess there would be still the manual step of adding the scripts from the data source?

**Reply by abhi1693** [reply]
You can automate that too using initializers during the bootup process of your pod

**Reply by rasanentimo** [reply]
Actually, I mean if I add a new script/file to the git repo, I guess there's still a manual step to add new script(s) from the data source on an already running pod?
